File: examples/BusinessHoursDemo.java
package com.cisco.ccbu.cce.unifiedconfig.toolkit.examples;

import com.cisco.ccbu.cce.unifiedconfig.toolkit.ApiException;
import com.cisco.ccbu.cce.unifiedconfig.toolkit.RESTClient;
import com.cisco.ccbu.cce.unifiedconfig.toolkit.ReferenceBean;
import com.cisco.ccbu.cce.unifiedconfig.toolkit.bean.*;

import java.util.ArrayList;
import java.util.List;

public class BusinessHoursDemo {

    private static RESTClient restClient = null;
    private static final String BUSINESS_HOUR_NAME_PREFIX = "DemoBH";

    /**
     * Validates the arguments
     */
    public static void main(String[] args) {
        // Create a new RESTClient object
        //args[0] - IP of AW-HDS
        //args[1] - user name(with domain)
        //args[2] - password
        if(args.length < 3) {
            System.out.println("BusinessHourDemoDemo requires 3 parameters: hostname username (with @domain) password");
            System.exit(0);
        }

        restClient = new RESTClient(args[0], args[1], args[2]);
        demoBusinessHours(restClient);
    }

    private static void demoBusinessHours(RESTClient restClient) {
        BusinessHour businessHour1 = new BusinessHour();
        BusinessHour businessHour2 = new BusinessHour();
        try {
            // -- Create a business hour
            businessHour1 = createAndGetBusinessHourConfig(restClient, BUSINESS_HOUR_NAME_PREFIX+1);

            // -- Update the business hour
            updateAndVerifyBusinessHourConfig(restClient, businessHour1);

            // -- Create 1 more business hour
            businessHour2 = createAndGetBusinessHourConfig(restClient, BUSINESS_HOUR_NAME_PREFIX+2);

            // Get the list of the Business hours created
            getBusinessHoursList();
        } catch (ApiException e){
            System.out.println(e.getErrors());
        }
        finally {
            // *** Finally Delete everything
            if(businessHour1 != null && businessHour1.getRefURL() != null) {
                restClient.delete(businessHour1.getRefURL());
                System.out.println("Deleted Business Hour: " + businessHour1.getRefURL());
            }
            if(businessHour2 != null && businessHour2.getRefURL() != null) {
                restClient.delete(businessHour2.getRefURL());
                System.out.println("Deleted Business Hour: " + businessHour2.getRefURL());
            }
            getBusinessHoursList();
        }
    }

    private static void getBusinessHoursList() {
        try {
            BusinessHour.BusinessHourList businessHourList = restClient.getList(BusinessHour.BusinessHourList.class, BUSINESS_HOUR_NAME_PREFIX);
            System.out.println("Number of Business Hours retrieved on LIST: " + businessHourList.getItems().size());
        } catch(ApiException e) {
            System.out.println(e.getErrors());
        }
    }

    private static BusinessHour createAndGetBusinessHourConfig(RESTClient restClient, String bhName) {
        BusinessHour businessHour = restClient.createAndGetBean(getBusinessHourForCreate(bhName));
        System.out.println("Business Hour created: " + businessHour.getRefURL());
        return businessHour;
    }

    /**
     * create a business Hour object for Create.
     *
     * @return
     */
    private static BusinessHour getBusinessHourForCreate(String bhName) {
        BusinessHour bh = new BusinessHour();
        bh.setName(bhName);
        bh.setType(1);
        bh.setDescription("This is for Business Hours Demo");

        BusinessHourConfiguredStatus businessHourConfiguredStatus = new BusinessHourConfiguredStatus();
        businessHourConfiguredStatus.setStatus(1);

        ReferenceBean businessHourStatusReasonRef = new ReferenceBean();
        businessHourStatusReasonRef.setRefURL("/unifiedconfig/config/businesshourstatusreason/5000");
        businessHourConfiguredStatus.setStatusReason(businessHourStatusReasonRef);

        bh.setConfiguredStatus(businessHourConfiguredStatus);

        ReferenceBean timeZoneLocationRef = new ReferenceBean();
        timeZoneLocationRef.setRefURL("/unifiedconfig/config/timezone/v2/5000");

        bh.setTimezone(timeZoneLocationRef);

        List<SpecialDaySchedule> specialDayScheduleList = new ArrayList<>();

        SpecialDaySchedule specialDaySchedule = new SpecialDaySchedule();
        specialDaySchedule.setDate("04-02-3019");
        specialDaySchedule.setStartTime("03:26");
        specialDaySchedule.setEndTime("04:16");
        specialDaySchedule.setDescription("Test");
        specialDaySchedule.setStatus(1);
        specialDaySchedule.setStatusReason(businessHourStatusReasonRef);
        specialDayScheduleList.add(specialDaySchedule);
        bh.setSpecialDaySchedules(specialDayScheduleList);

        List<WeekDaySchedule> weekDayScheduleList = new ArrayList<>();

        WeekDaySchedule weekDaySchedule = new WeekDaySchedule();
        weekDaySchedule.setDayOfWeek(1);
        weekDaySchedule.setStartTime("20:20");
        weekDaySchedule.setEndTime("23:00");
        weekDayScheduleList.add(weekDaySchedule);

        bh.setWeekDaySchedules(weekDayScheduleList);

        return bh;
    }

    /**
     * Update a business Hour object and verify that it's updated.
     *
     * @return
     */
    private static void updateAndVerifyBusinessHourConfig(RESTClient restClient, BusinessHour bh) {
        System.out.println("Business Hour time zone before update: " + bh.getTimezone().getRefURL());

        ReferenceBean timeZoneLocationRef = new ReferenceBean();
        timeZoneLocationRef.setRefURL("/unifiedconfig/config/timezone/v2/5001");

        bh.setTimezone(timeZoneLocationRef);
        BusinessHour updatedBH = restClient.updateAndGetBean(bh);
        System.out.println("Business Hour updated: " + updatedBH.getRefURL());
        System.out.println("Business Hour time zone after update: " + bh.getTimezone().getRefURL());
    }
}

File: CiscoPCCE.Toolkit/Examples/BusinessHoursDemo.cs
using CiscoPCCE.Toolkit.Bean;
using CiscoPCCE.Toolkit.Sdk;
using CiscoPCCE.Toolkit.Sdk.Helpers;

namespace CiscoPCCE.Toolkit.Examples
{
    /// <summary>
    /// Example code for Business Hours operations.
    /// </summary>
    public class BusinessHoursDemo
    {
        private static RestClient? restClient = null;
        private const string BusinessHourNamePrefix = "DemoBH";

        public static async Task Main(string[] args)
        {
            // Create a new RESTClient object
            //args[0] - IP of AW-HDS
            //args[1] - user name(with domain)
            //args[2] - password
            if (args.Length < 3)
            {
                Console.WriteLine("BusinessHoursDemo requires 3 parameters: hostname username (with @domain) password");
                Environment.Exit(0);
            }

            restClient = new RestClient(args[0], args[1], args[2]);
            await DemoBusinessHoursAsync(restClient);
        }

        private static async Task DemoBusinessHoursAsync(RestClient restClient)
        {
            BusinessHour? businessHour1 = null;
            BusinessHour? businessHour2 = null;

            try
            {
                // -- Create a business hour
                businessHour1 = await CreateAndGetBusinessHourConfigAsync(restClient, BusinessHourNamePrefix + "1");

                // -- Update the business hour
                await UpdateAndVerifyBusinessHourConfigAsync(restClient, businessHour1);

                // -- Create 1 more business hour
                businessHour2 = await CreateAndGetBusinessHourConfigAsync(restClient, BusinessHourNamePrefix + "2");

                // Get the list of the Business hours created
                await GetBusinessHoursListAsync(restClient);
            }
            catch (ApiException e)
            {
                Console.WriteLine(e.Errors);
            }
            finally
            {
                // *** Finally Delete everything
                if (businessHour1?.RefURL != null)
                {
                    await restClient.DeleteAsync(businessHour1.RefURL);
                    Console.WriteLine($"Deleted Business Hour: {businessHour1.RefURL}");
                }
                if (businessHour2?.RefURL != null)
                {
                    await restClient.DeleteAsync(businessHour2.RefURL);
                    Console.WriteLine($"Deleted Business Hour: {businessHour2.RefURL}");
                }
                await GetBusinessHoursListAsync(restClient);
            }
        }

        private static async Task GetBusinessHoursListAsync(RestClient restClient)
        {
            try
            {
                var businessHourList = await restClient.GetListAsync<BusinessHourList>(BusinessHourNamePrefix);
                Console.WriteLine($"Number of Business Hours retrieved on LIST: {businessHourList?.Items?.Count ?? 0}");
            }
            catch (ApiException e)
            {
                Console.WriteLine(e.Errors);
            }
        }

        private static async Task<BusinessHour?> CreateAndGetBusinessHourConfigAsync(RestClient restClient, string bhName)
        {
            var bean = GetBusinessHourForCreate(bhName);
            var refUrl = await restClient.CreateAndGetAsync(bean);
            BusinessHour? businessHour = null;
            if (refUrl != null)
            {
                businessHour = await restClient.GetAsync<BusinessHour>(refUrl);
                Console.WriteLine($"Business Hour created: {businessHour?.RefURL}");
            }
            return businessHour;
        }

        /// <summary>
        /// Create a business Hour object for Create.
        /// </summary>
        private static BusinessHour GetBusinessHourForCreate(string bhName)
        {
            var businessHourConfiguredStatus = new BusinessHourConfiguredStatus
            {
                Status = 1,
                StatusReason = new ReferenceBean
                {
                    RefURL = "/unifiedconfig/config/businesshourstatusreason/5000"
                },
                Department = new ReferenceBean()
            };

            var timeZoneLocationRef = new ReferenceBean
            {
                RefURL = "/unifiedconfig/config/timezone/v2/5000"
            };

            var bh = new BusinessHour
            {
                Name = bhName,
                Type = 1,
                Description = "This is for Business Hours Demo",
                ConfiguredStatus = businessHourConfiguredStatus,
                Timezone = timeZoneLocationRef,
                Department = new ReferenceBean()
            };

            var specialDayScheduleList = new List<SpecialDaySchedule>();

            var specialDaySchedule = new SpecialDaySchedule
            {
                Date = "04-02-3019",
                StartTime = "03:26",
                EndTime = "04:16",
                Description = "Test",
                Status = 1,
                StatusReason = new ReferenceBean
                {
                    RefURL = "/unifiedconfig/config/businesshourstatusreason/5000"
                },
                Department = new ReferenceBean()
            };
            specialDayScheduleList.Add(specialDaySchedule);
            bh.SpecialDaySchedules = specialDayScheduleList;

            var weekDayScheduleList = new List<WeekDaySchedule>();

            var weekDaySchedule = new WeekDaySchedule
            {
                DayOfWeek = 1,
                StartTime = "20:20",
                EndTime = "23:00",
                Department = new ReferenceBean()
            };
            weekDayScheduleList.Add(weekDaySchedule);

            bh.WeekDaySchedules = weekDayScheduleList;

            return bh;
        }

        /// <summary>
        /// Update a business Hour object and verify that it's updated.
        /// </summary>
        private static async Task UpdateAndVerifyBusinessHourConfigAsync(RestClient restClient, BusinessHour bh)
        {
            Console.WriteLine($"Business Hour time zone before update: {bh.Timezone?.RefURL}");

            var timeZoneLocationRef = new ReferenceBean
            {
                RefURL = "/unifiedconfig/config/timezone/v2/5001"
            };

            bh.Timezone = timeZoneLocationRef;
            var updatedBH = await restClient.UpdateAndGetBeanAsync(bh);
            Console.WriteLine($"Business Hour updated: {updatedBH?.RefURL}");
            Console.WriteLine($"Business Hour time zone after update: {bh.Timezone?.RefURL}");
        }
    }
}
